### PR TITLE
5373: (doc) DROP TYPE and DROP INDEX

### PR DESCRIPTION
**Pull request by msudberg**
Special concerns: Am I correct that DROP INDEX converts a materialized view to a nonmaterialized view rather than deleting it per se?

Also to confirm scope:

The following commands don't seem to support CASCADE | RESTRICT:

DROP DATABASE
DROP ROLE
DROP SINK
DROP USER

The following are already documented:

DROP SCHEMA
DROP SOURCE
DROP VIEW

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/5956)

### CODE REVIEW

File: doc/user/content/sql/drop-index.md [hunk 1]
```diff
@@ -42,8 +45,12 @@ If you drop the primary index and want to recreate it:
 Alternatively, you can also [`CREATE INDEX`](../create-index) to materialize any
 view.
 
+Dropping the only index on a materialized view does not remove the view entirely; it merely transforms it to a non-materialized view.
+
 ## Examples
 
+### Drop indexes (no dependencies)
```
Comment (JLDLaughlin) [inline doc/user/content/sql/drop-index.md:52]: Maybe `Drop indexes without dependent objects` and `Drop indexes with dependent objects` below to make these more symmetric?

File: doc/user/content/sql/drop-index.md [hunk 2]
```diff
@@ -42,8 +45,12 @@ If you drop the primary index and want to recreate it:
 Alternatively, you can also [`CREATE INDEX`](../create-index) to materialize any
 view.
 
+Dropping the only index on a materialized view does not remove the view entirely; it merely transforms it to a non-materialized view.
+
 ## Examples
```
Comment (JLDLaughlin) [inline doc/user/content/sql/drop-index.md:50]: Could you add examples for `IF EXISTS` and `RESTRICT`?

File: doc/user/content/sql/drop-table.md [hunk 3]
```diff
@@ -19,30 +19,68 @@ statements. `DROP TABLE` removes tables from your Materialize instance.
...
+`IF EXISTS`  | Do not issue an error if the named table doesn't exist.
+`RESTRICT`  | Don't remove the table if any objects depend on it. _(Default.)_
+  |
 ## Examples
 
-### Dropping a table
-
-We can create a table `t` and verify that it was created with the following
-statements:
+### Drop a table (no dependencies)
```
Comment (JLDLaughlin) [inline doc/user/content/sql/drop-table.md:29]: Same thought as for indexes! Happy with any format, as long as it is consistent across pages.

File: doc/user/content/sql/drop-table.md [hunk 4]
```diff
@@ -19,30 +19,68 @@ statements. `DROP TABLE` removes tables from your Materialize instance.
...
 ------|-----
-**IF EXISTS** | Do not return an error if the named table does not exist.
-_table&lowbar;name_ | The name of the table to drop.
-
+_data_type_name_ | The name of the table to remove.
+`CASCADE` | Automatically removes any objects that depend on the table, as well as the table.
+`IF EXISTS`  | Do not issue an error if the named table doesn't exist.
+`RESTRICT`  | Don't remove the table if any objects depend on it. _(Default.)_
+  |
 ## Examples
```
Comment (JLDLaughlin) [inline doc/user/content/sql/drop-table.md:27]: Can you add `IF EXISTS` and `RESTRICT` examples here, too?

File: doc/user/content/sql/drop-type.md [hunk 5]
```diff
@@ -17,6 +17,7 @@ menu:
 Field | Use
 ------|-----
 _data_type_name_ | The name of the type to remove.
+
```
Comment (JLDLaughlin) [inline doc/user/content/sql/drop-type.md:20]: Errant newline?

**Review by JLDLaughlin — COMMENTED** [review]
I had a few suggestions! I also am happy to review (mostly for technical correctness), but I think @sploiselle might have more opinions on copy? I'm less aware of consistency across pages, etc.

File: doc/user/content/sql/drop-index.md [hunk 6]
```diff
@@ -16,6 +16,9 @@ do not have any indexes.)
 Field | Use
 ------|-----
 _view&lowbar;name_ | The name of the index you want to remove.
+`CASCADE` | Automatically removes any objects that depend on the index, as well as the index.
+`IF EXISTS`  |  Do not issue an error if the named index doesn't exist.
+`RESTRICT`  |  Remove the index, converting the materialized view to a non-materialized view. _(Default)._
```
Comment (sploiselle) [inline doc/user/content/sql/drop-index.md:21]: This is the wrong description here; the other table has the correct description of `RESTRICT`.

File: doc/user/content/sql/drop-index.md [hunk 7]
```diff
@@ -27,8 +30,8 @@ VIEW`](../create-view).
 
 ### Primary indexes
 
-By default, materialized views only have one index, which we call the "primary
-index," and stores the result set of the materialized view's embedded `SELECT`
+By default, materialized views only have one index, called the "primary
```
Comment (sploiselle) [inline doc/user/content/sql/drop-index.md:33]: The term "primary index" has a distinct meaning in the world of databases, and our "primary indexes" do not work like primary indexes do in traditional databases, hence the oddly phrased "which we call." I was trying to intimate at this fact, but perhaps wasn't explicit enough.

File: doc/user/content/sql/drop-index.md [hunk 8]
```diff
@@ -69,6 +76,33 @@ SHOW INDEXES FROM q01;
...
+```
+```nofmt
++------------------------+--------------------------------+-----
+| View                   | Key_name                       | ...
+|------------------------+--------------------------------+----
+| materialize.public.q01 | materialize.public.q01_geo_idx | ...
++------------------------+--------------------------------+-----
+```
+```sql
+DROP INDEX materialize.public.q01_geo_idx CASCADE;
```
Comment (sploiselle) [inline doc/user/content/sql/drop-index.md:103]: We don't typically include the full qualification in these examples, which is a potentially instructive point that index names are referrable by the same rules as all other catalog objects.

File: doc/user/content/sql/drop-table.md [hunk 9]
```diff
@@ -19,30 +19,68 @@ statements. `DROP TABLE` removes tables from your Materialize instance.
 
 Field | Use
 ------|-----
-**IF EXISTS** | Do not return an error if the named table does not exist.
-_table&lowbar;name_ | The name of the table to drop.
-
+_data_type_name_ | The name of the table to remove.
```
Comment (sploiselle) [inline doc/user/content/sql/drop-table.md:22]: the `Field` is wrong here either here or in the railroad diagram

File: doc/user/content/sql/drop-table.md [hunk 10]
```diff
@@ -19,30 +19,68 @@ statements. `DROP TABLE` removes tables from your Materialize instance.
...
+`CASCADE` | Automatically removes any objects that depend on the table, as well as the table.
```
Comment (sploiselle) [inline doc/user/content/sql/drop-table.md:23]: The verb tense here differs from the others.

File: doc/user/content/sql/drop-index.md (hunk 7 above)
Comment (msudberg) [reply doc/user/content/sql/drop-index.md:33]: @sploiselle - Can you give me a quick summary of the difference between Materialize primary indexes and trad db primary indexes? Basically, what user expectations will we confound?

Comment (sploiselle) [reply doc/user/content/sql/drop-index.md:33]: The term primary index in a traditional database comes freighted with meaning, especially for folks familiar with that database's behavior regarding indices. For example, the primary index is often thought of as the index that enforces the invariant that provides a table's primary key (i.e. the data in the columns in the primary is a unique set in each row, which also implies that it is not null). In some systems, the primary key also describes how to store the canonical version of the data on disk (or in memory)––i.e. all other indices on the table are derivatives of the primary index.

What we call our "primary index" is just the first index created on a view. You can drop it and, if you have another covering index on the view (covering in this context just means that the index includes all of the columns), I don't believe there are any untoward effects. However, if you drop the "primary index" and do not have another covering index (if it's even possible to have a non-covering index in this case?), you can no longer query the view (in the general case), or at least cannot query any column not included in an index. In a traditional RDBMS, if you drop the primary index, you can still query the table, it will just be very slow––i.e. in the general case, the RDBMS will essentially create a "new" primary index whose primary key is the empty set and each row is essentially just a bag of bytes you have to sift through "manually" to do any kind of sorting.

There are also things like this that are just very dramatically different from the kind of behavior one expects out of primary indices in a traditional database:

> You cannot exclude any columns from being in the index’s “value” set. For example, if your view embedded `SELECT a, b, c..`., all indexes will contain `{a, b, c}` as their values.

In a traditional RDBMS, you would only include the on-indexed columns in the value set.

All of this is "basically" true, but there are certainly edge cases that this glosses over. Indices in databases are an incredibly intricate topic with few universal truths because each RDBMS makes slightly different tradeoffs with how they handle indices to achieve some desirable set of performance characteristics.

File: doc/user/content/sql/drop-index.md (hunk 6 above)
Comment (msudberg) [reply doc/user/content/sql/drop-index.md:21]: @sploiselle @JLDLaughlin This is what I get when I try it:

```
SHOW FULL VIEWS;
     name      | type | materialized 
---------------+------+--------------
 t2_view       | user | t
 t_view        | user | t
```

`DROP INDEX t_view_primary_idx;`

```
SHOW FULL VIEWS;
     name      | type | materialized 
---------------+------+--------------
 t2_view       | user | t
 t_view        | user | f
```

Is this not converting the materialized view to a nonmaterialized view instead of removing it?

Comment (sploiselle) [reply doc/user/content/sql/drop-index.md:21]: This comment was on the `RESTRICT` keyword, but the example you're providing elides it from the command. If you re-run the command using `RESTRICT` does it still drop the index?

File: doc/user/content/sql/drop-index.md (hunk 7 above)
Comment (msudberg) [reply doc/user/content/sql/drop-index.md:33]: @sploiselle Thanks! Does this work?

> By default, materialized views only have one index, which we call the "primary
index," and which stores the result set of the materialized view's embedded `SELECT`
statement. You can identify the primary index by its name, which follows the
format `<view name>_primary_idx`.
>
> The materialized view primary index doesn't follow the constraints for typical for primary indexes in traditional databases; it is simply the first index created on a view. However, if the view doesn't have another index that covers all of the columns, you will not be able to query any columns not included in an existing index.

File: doc/user/content/sql/drop-index.md (hunk 6 above)
Comment (sploiselle) [reply doc/user/content/sql/drop-index.md:21]: @msudberg You're right––it should be implied. This looks like a bug.

Comment (sploiselle) [reply doc/user/content/sql/drop-index.md:21]: Ah, ok––it looks like we shouldn't have `RESTRICT` (or `CASCADE`) as an option on indexes yet because we don't have a notion of "dependency" for indexes afaict. The materialized view does not depend on the index (instead, the relationship goes the other way: the index depends on the view, though dropping the view doesn't require an explicit `CASCADE` to drop the index for ergonomic reasons). Something that would depend on the index would be something like a constraint, e.g. `NOT NULL`.

For future reference, a better place to pull the supported syntax from would be one of the test drive test files; in this case `test/testdrive/createdrop.td`. You can find the appropriate file by looking for the command you want (usually in all caps) in `*.td` files using e.g. VSCode (I think you use Atom, but there should be a similar search filtering option there).

One could make an argument that we shouldn't accept either cascade or restrict options because they're both definitely no-ops and it's confusing as to what to expect, but one could also make an argument that this is working as intended and obeys both the requested `CASCADE` and `RESTRICT` behavior.

@JLDLaughlin Does this track for you, as well?

File: doc/user/content/sql/drop-index.md [hunk 11]
```diff
@@ -16,6 +16,9 @@ do not have any indexes.)
 Field | Use
 ------|-----
 _view&lowbar;name_ | The name of the index you want to remove.
+`CASCADE` | Automatically removes any objects that depend on the index, as well as the index.
```
Comment (sploiselle) [inline doc/user/content/sql/drop-index.md:19]: Just flagging that we should revisit this inclusion

File: doc/user/content/sql/drop-table.md [hunk 12]
```diff
@@ -19,30 +19,68 @@ statements. `DROP TABLE` removes tables from your Materialize instance.
...
+`IF EXISTS`  | Do not issue an error if the named table doesn't exist.
+`RESTRICT`  | Don't remove the table if any objects depend on it. _(Default.)_
```
Comment (sploiselle) [inline doc/user/content/sql/drop-table.md:25]: this should be "any non-index objects"; we'll drop the table and its indexes even under `RESTRICT`.

File: doc/user/content/sql/drop-index.md (hunk 7 above)
Comment (sploiselle) [reply doc/user/content/sql/drop-index.md:33]: The second paragraph needs some TLC copy-edit-wise, but I believe captures the gist. Ping me once you've cleaned it up.

File: doc/user/content/sql/drop-index.md (hunk 6 above)
Comment (JLDLaughlin) [reply doc/user/content/sql/drop-index.md:21]: Yes, it does! Thank you both for digging into this. 

@sploiselle - updating materialized views to depend on indexes would... I'm not sure about all the ramifications, but create a cyclical dependency that would probably just require a lot of special handling. Do you think this is worth pursuing? Or do you think the current `RESTRICT` and `CASCADE` behavior (however weird) is sufficient?

Comment (msudberg) [reply doc/user/content/sql/drop-index.md:21]: @sploiselle - I actually did search using Atom! `DROP INDEX CASCADE` shows up in materialize/test/sqllogictest/cockroach/alter_table.slt, but I don't see anything relevant in my local version of testdrive.td.

@JLDLaughlin - Is it actually correct to say that `DROP INDEX` converts materialized views to non-materialized views, or should the definition just be the usual "Removes materialized views"?

Comment (sploiselle) [reply doc/user/content/sql/drop-index.md:21]: @msudberg Ah––there are, of course, a million contingencies. A lot of features in the `materialize/test/sqllogictest/cockroach/` directory are not supported. The set of "canonical" tests is in `materialize/test/sqllogictest`.

I realize you asked this to Jessica, but maybe I can help get you unblocked: `DROP INDEX` removes the index, which converts a materialized view into a non-materialized view iff there are no indices remaining after this index gets dropped. `DROP INDEX` doesn't/can't drop the view it's on, so it would be a little confusing to say it removes the materialized view because the view itself remains, just in unmaterialized form.

Comment (msudberg) [reply doc/user/content/sql/drop-index.md:21]: @sploiselle, @JLDLaughlin - What is the final call on including or omitting `RESTRICT` and `CASCADE`?

Comment (sploiselle) [reply doc/user/content/sql/drop-index.md:21]: I would suggest mentioning that it's valid syntax but is a no-op because nothing in Materialize can depend on indices. I don't think there's a clear answer around whether to e.g. include it in the railroad diagram, so defer to your judgment on the presentation.

Comment (msudberg) [reply doc/user/content/sql/drop-index.md:21]: I think if CASCADE, RESTRICT, and no switch all do the exact same thing, we might as well omit CASCADE and RESTRICT and prevent the confusion. We do have other DROP commands that don't support those switches.

Comment (sploiselle) [reply doc/user/content/sql/drop-index.md:21]: Helping prevent the confusion from a documentation perspective would be to explain the contingency where users might encounter it, rather than papering over it by omitting the keyword and hoping no one types it in. There is definitely engineering work we could do to improve this UX, but a note in the docs takes a much less significant amount of effort and is a good intermediary solution until the engineering team can take on more of this polish work.

**Comment by msudberg** [comment]
@sploiselle @JLDLaughlin - Ready for next round of review.

File: doc/user/content/sql/drop-index.md [hunk 13]
```diff
@@ -6,33 +6,39 @@ menu:
...
 By default, materialized views only have one index, which we call the "primary
-index," and stores the result set of the materialized view's embedded `SELECT`
+index", and which stores the result set of the materialized view's embedded `SELECT`
 statement. You can identify the primary index by its name, which follows the
 format `<view name>_primary_idx`.
 
-If you drop the primary index and want to recreate it:
+This primary index doesn't follow the constraints for typical for primary indexes in traditional databases; it is simply the first index created on a view. Please keep in mind the following constraints:
+
+* If you remove the primary index, you will not be able to query any columns that are not included in another index for the view.
```
[the remark this answers is not in the document]
Comment (sploiselle) [reply doc/user/content/sql/drop-index.md:37]: here and elsewhere

File: doc/user/content/sql/drop-schema.md [hunk 14]
```diff
@@ -26,20 +26,42 @@ Before you can drop a schema, you must [drop all sources](../drop-source) and
...
 ```
 ```sql
-DROP SCHEMA my_schema CASCADE
+DROP SCHEMA my_schema CASCADE;
+```
+
+### Remove a schema only if it has no dependent objects
+
+```sql
+DROP SCHEMA my_schema;
```
Comment (sploiselle) [inline doc/user/content/sql/drop-schema.md:54]: We typically don't include the tautological outputs of commands, but we do include errors because the provide meaningful context that help users understand the examples.

File: doc/user/content/sql/drop-table.md [hunk 15]
```diff
@@ -19,31 +19,85 @@ statements. `DROP TABLE` removes tables from your Materialize instance.
...
+DROP TABLE t;
+```
+### Remove a table with dependent objects
+
+Create a table *t*:
+
+```sql
+CREATE TABLE t (a int, b text NOT NULL);
+INSERT INTO t VALUES (1, 'yes'), (2, 'no'), (3, 'maybe');
+SELECT * FROM t;
```
Comment (sploiselle) [inline doc/user/content/sql/drop-table.md:55]: These examples can be a little easier for users to parse if you include an `ORDER BY` clause, but not strictly necessary.

File: doc/user/content/sql/drop-table.md [hunk 16]
```diff
@@ -19,31 +19,85 @@ statements. `DROP TABLE` removes tables from your Materialize instance.
...
+
+Create a materialized view from *t*:
+
+```sql
+CREATE MATERIALIZED VIEW t_view AS SELECT sum(a) AS sum FROM t;
+SHOW MATERIALIZED VIEWS;
+```
+```
+    name
+---------------
```
Comment (sploiselle) [inline doc/user/content/sql/drop-table.md:74]: this formatting looks off; the way that the dashes center under the column name doesn't do this in the output itself

File: doc/user/content/sql/drop-type.md [hunk 17]
```diff
@@ -16,14 +16,14 @@ menu:
 
 Field | Use
 ------|-----
+`IF EXISTS`  | Do not return an error if the named type doesn't exist.
 _data_type_name_ | The name of the type to remove.
-`CASCADE` | Automatically removes any objects that depend on the type, such as tables or other types.
-`IF EXISTS`  |  Do not issue an error if the named type doesn't exist.
-`RESTRICT`  |  Do not remove the type if any other objects depend on it. _(Default.)_
+`CASCADE` | Remove the type and its dependent objects, such as tables or other types.
+`RESTRICT`  |  Don't remove the type if any objects depend on it. _(Default.)_
```
Comment (sploiselle) [inline doc/user/content/sql/drop-type.md:22]: errant double space after `RESTRICT`

File: doc/user/content/sql/drop-index.md [hunk 18]
```diff
@@ -65,8 +73,26 @@ SHOW INDEXES FROM q01;
 | materialize.public.q01 | materialize.public.q01_geo_idx | ...
 +------------------------+--------------------------------+-----
 ```
+
+You can use the unqualified index name (`q01_geo_idx`) rather the fully qualified name (`materialize.public.q01_geo_idx`).
+
+
```
[the remark this answers is not in the document]
Comment (sploiselle) [reply doc/user/content/sql/drop-index.md:79]: Looking at some of the other examples, maybe you mean to make this an exhaustive set of commands that perform these actions? I think that could be fine but needs more context that this is what you're showing using some kind of lead in and maybe an unordered list to show parallel construction? I think that this is slightly overwrought, though, and you'd be just fine with only the vanilla command.

File: doc/user/content/sql/drop-type.md [hunk 19]
```diff
@@ -84,6 +84,22 @@ SELECT * FROM t;
 ERROR:  unknown catalog item 't'
 ```
 
+### Remove a type only if it has no dependent objects
```
Comment (sploiselle) [inline doc/user/content/sql/drop-type.md:87]: Just seconding the suggestion here to make clear that you are not showing sequential commands you could run like the examples elsewhere.

File: doc/user/content/sql/show-types.md [hunk 20]
```diff
@@ -22,22 +22,17 @@ Field | Use
 ### Show custom data types
 
 ```sql
-CREATE TYPE int4_list AS LIST (element_type = int4);
```
Comment (sploiselle) [inline doc/user/content/sql/show-types.md:25]: Why did you remove this?

**Review by sploiselle — CHANGES_REQUESTED** [review]

File: doc/user/content/sql/drop-index.md [hunk 21]
```diff
@@ -6,33 +6,39 @@ menu:
...
 
-If you drop the only index on a materialized view, it becomes a non-materialized
-view. For more details on non-materialized views, see [`CREATE
-VIEW`](../create-view).
-
 ### Primary indexes
 
 By default, materialized views only have one index, which we call the "primary
-index," and stores the result set of the materialized view's embedded `SELECT`
+index", and which stores the result set of the materialized view's embedded `SELECT`
```
[the remark this answers is not in the document]
Comment (msudberg) [reply doc/user/content/sql/drop-index.md:31]: I think it's harder to parse the sentence without the "and", because  then you have two dependent clauses starting with "which" in a row.

File: doc/user/content/sql/drop-index.md (hunk 18 above)
[the remark this answers is not in the document]
Comment (msudberg) [reply doc/user/content/sql/drop-index.md:79]: Jessica asked for examples of CASCADE and RESTRICT. Since Markdown doesn't play nice with codeblocks and unordered lists, and codephrases aren't consistent with what we do elsewhere, I've add an explanatory note.

[the remark this answers is not in the document]
Comment (sploiselle) [reply doc/user/content/sql/drop-index.md:80]: Markdown does play nicely with code blocks and unordered lists. If you can't get something to work, you should ask for help.
